cmd/common/registry.go

```diff
@@ -0,0 +1,340 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/frontends/legacy"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	columnhelpers "github.com/inspektor-gadget/inspektor-gadget/internal/column-helpers"
+	gadgetrunner "github.com/inspektor-gadget/inspektor-gadget/internal/gadget-runner"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/runtime"
+	cols "github.com/inspektor-gadget/inspektor-gadget/pkg/columns"
+	gadgetregistry "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-registry"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+// AddCommandsFromRegistry adds all gadgets known by the registry as cobra commands as a subcommand to their categories
+func AddCommandsFromRegistry(rootCmd *cobra.Command, runtime runtime.Runtime, columnFilters []cols.ColumnFilter) {
+	runtimeParams := runtime.Params()
+
+	// Build lookup
+	lookup := make(map[string]*cobra.Command)
+	for _, cmd := range rootCmd.Commands() {
+		lookup[cmd.Name()] = cmd
+	}
+
+	// Add runtime flags
+	addFlags(rootCmd, runtimeParams)
+
+	// Add operator global flags
+	operatorsParamsCollection := operators.OperatorsParamsCollection()
+	for _, operatorParams := range operatorsParamsCollection {
+		addFlags(rootCmd, operatorParams)
+	}
+
+	// Add all known gadgets to cobra in their respective categories
+	categories := gadgets.GetCategories()
+	for _, gadget := range gadgetregistry.GetGadgets() {
+		cmd, ok := lookup[gadget.Category()]
+		if !ok {
+			// Category not found, add it
+			categoryDescription, ok := categories[gadget.Category()]
+			if !ok {
+				panic(fmt.Errorf("category unknown: %q", gadget.Category()))
+			}
+			cmd = &cobra.Command{
+				Use:   gadget.Category(),
+				Short: categoryDescription,
+			}
+			rootCmd.AddCommand(cmd)
+			lookup[gadget.Category()] = cmd
+		}
+		cmd.AddCommand(buildCommandFromGadget(gadget, columnFilters, runtime, runtimeParams, operatorsParamsCollection))
```

Comment:
Maybe I misunderstood, but I think this relates to [this discussion](https://github.com/inspektor-gadget/inspektor-gadget/pull/1281#discussion_r1086711634).

`Params()` always returns a fresh copy and doesn't keep the values inside runtime/operator/gadget - to be able to use all components with different parameter values. In case of runtime and the global operator params, this is done inside `AddCommandsFromRegistry` to only add them once.